lib/experimental/html/class-wp-html-tag-processor.php

```diff
@@ -1412,6 +1412,40 @@ public function get_attribute( $name ) {
 		return html_entity_decode( $raw_value );
 	}
 
+	/**
+	 * Returns the names of all attributes matching a given prefix in the currently-opened tag.
+	 *
+	 * Example:
+	 * <code>
+	 *     $p = new WP_HTML_Tag_Processor( '<div data-enabled class="test" data-test-id="14">Test</div>' );
+	 *     $p->next_tag( [ 'class_name' => 'test' ] ) === true;
+	 *     $p->get_attribute_names_with_prefix() === array( 'data-enabled', 'data-test-id' );
+	 *
+	 *     $p->next_tag( [] ) === false;
+	 *     $p->get_attribute_names_with_prefix() === null;
+	 * </code>
+	 *
+	 * @since 6.2.0
+	 *
+	 * @param string $prefix Prefix of requested attribute names.
+	 * @return array|null List of attribute names, or `null` if not at a tag.
+	 */
+	function get_attribute_names_with_prefix( $prefix ) {
+		if ( $this->is_closing_tag || null === $this->tag_name_starts_at ) {
+			return null;
+		}
+
+		$comparable = strtolower( $prefix );
+
+		$matches = array_filter(
+			array_keys( $this->attributes ),
+			function( $attr ) use ( $comparable ) {
+				return str_starts_with( $attr, $comparable );
+			}
+		);
+		return array_values( $matches );
```

Comment:
> note too that we can't use `str_starts_with()` as that appeared in PHP 8

It's polyfiled -- and used -- in Core though: https://github.com/WordPress/wordpress-develop/blob/acbbee8a1199f9ce70f1d3bbcc9cb80e32047e6c/src/wp-includes/compat.php#L445-L465 😊

Reply:
> I believe we would do better to avoid the array functions and inline the iteration here, particularly because we're talking about running this code on every tag in every block that supports WP directives.

When writing this code, I also wasn't too happy I had to run `array_values()` after `array_filter()` (and `array_keys()`. If I'm not mistaken, the complexity is:
1. `array_keys`: `O(n)`, where `n` is the length of `$this->attributes`
2. `array_filter`: same
3. `array_value`: `O(m)`, where `m` is the length of attributes matching the prefix (thus `m <= n`)

I've found a way (a4b10719aa) that lets me get rid of one of those iterations, resulting in:

1. `array_filter`: `O(n)`
2. `array_keys`: `O(m)`

I think that this should be equivalent to, or even slightly better than (because `n <= m`), the "manual" `foreach` iteration:

1. `array_keys`: `O(n)`
2. `foreach`: `O(n)`

Reply:
Quick test on `single-page.html` grabbing the names of every attribute of every tag (282,414 tags, 214,741 attributes) - running 30 rounds before measuring to prime any caches, measuring the successive 80 rounds, only measuring time to iterate tags, not the time instantiating the tag processor. I ran each test with my fans set to 100% speed to eliminate variation due to thermal issues. All times are in milliseconds; all memory numbers in bytes.

| Version | Mean | Median | StdDev | Min |
|---|---|---|---|---|
| `array_filter(USE_KEY)` | 698 | 694 | 9.72 | 684 |
| `foreach(array_keys)` | 638 | 636 | 6.62 | 630 |
| Don't get attribute names | 610 | 607 | 11.2 | 593 |

While this benchmark is somewhat of a non-normative example of content going through here, I think it gives a decent baseline for understanding the impact of variations of accomplishing a given goal inside PHP.

It looks like the `array_filter` is around 9%+ slower than `foreach`, despite the theoretical speedup. These ran on PHP 8.2.0.

Re-running with `get_attribute_names_with_prefix( 'data-' )` finds 12,662 such attributes.

| Version | Mean | Median | StdDev | Min | Additional Allocated Mem |
|---|---|---|---|---|---|
| `array_filter(USE_KEY)` | 683 | 679 | 8.80 | 670 | 538,546 |
| `foreach(array_keys)` | 628 | 625 | 9.96 | 616 | 537,978 |

So memory use doesn't appear to be impacted much, probably because the set of attributes for any given tag is likely to be small, but CPU usage remains noticeably higher with the `array_filter()`

Reply:
Ran some more tests with 80 pre-rounds and 800 measured rounds. This time I was curious about calling a polyfill function. The difference between `str_starts_with()` and `0 === strncmp()` seems negligible on PHP 8, but to simulate pre-PHP8 environments where we have to call a user-space function I used the following change.

the times are longer because I've moved to a more-controlled testing environment but it appears that the CPU is slower than my laptop (2.0 GHz vs. 4.6 GHz).

```php
return self::starts_with( $attr, $comparable );
```

and

```php
public static function starts_with( $s, $prefix ) {
	return str_starts_with( $s, $prefix );
}
```

| Version | Mean | Median | StdDev | Min |
|---|---|---|---|---|
| `str_starts_with` | 941 | 941 | 1.81 | 938 |
| `self::starts_with` | 943 | 943 | 1.12 | 940 |
| `0 === strncmp` | 936 | 935 | 2.06 | 932 |

In this benchmark it doesn't appear like the wrapping function is adding much overhead. It could be that PHP inlined it but I think the conclusion is that it's fine either way. `strncmp` is still faster, and at $N = 800$ there is actually strong statistical confidence that each of these is actually faster/slower than the other based on the means.

Interestingly if I change the wrapper to check for the presence of the `str_starts_with` function the overhead jumps another 10ms.

```php
public static function starts_with( $s, $prefix ) {
	if ( function_exists( 'str_starts_with' ) ) {
		return str_starts_with( $s, $prefix );
	} else {
		return 0 === strncmp( $s, $prefix, strlen( $prefix ) );
	}
}
```

This is still running in PHP 8 so only the first branch should ever be running there. This is still only adding 1.6% of overhead for the convenience over `0 === strncmp()`, so I don't think it matters much, but if it were me in this performance-critical code I'd probably go with the one that performs the same computation by doing less work.

Reply:
derp; should have looked at the polypill, and I see it's adding the polyfill once and never checking on every call which to use.

so particularly given that fact I think this convenience actually only adds less than 0.2% of overhead, something negligible here in PHP 8 at least, due to the user-space function call vs. a built-in.

Reply:
Thank you very much for the exhaustive analysis, much appreciated! (Curious what you used to get the timings so I can do more of that myself 😊)

Based on this, I'll change the code to use `foreach` (but continue to use `str_starts_with`) 😊